src/bundler/ESBuildBundler.js

```diff
@@ -0,0 +1,186 @@
+/*
+ * Copyright 2019 Adobe. All rights reserved.
+ * This file is licensed to you under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License. You may obtain a copy
+ * of the License at http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under
+ * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
+ * OF ANY KIND, either express or implied. See the License for the specific language
+ * governing permissions and limitations under the License.
+ */
+import { fileURLToPath } from 'url';
+import path from 'path';
+import fse from 'fs-extra';
+import * as esbuild from 'esbuild';
+import chalk from 'chalk-template';
+
+import processQueue from '@adobe/helix-shared-process-queue';
+
+import BaseBundler from './BaseBundler.js';
+
+// eslint-disable-next-line no-underscore-dangle
+const __dirname = path.resolve(fileURLToPath(import.meta.url), '..');
+
+/**
+ * Webpack based bundler
+ */
+export default class ESBuildBundler extends BaseBundler {
+  /**
+   * Create a new bundler.
+   *
+   * @param {import('../BaseConfig.js')} cfg base config
+   */
+  constructor(cfg) {
+    super(cfg);
+    this.arch = 'node';
+    this.type = 'esbuild';
+  }
+
+  async init() {
+    if (!this.cfg.esm) {
+      throw new Error('ESBuild bundler only supports ESM builds.');
+    }
+  }
+
+  async getESBuildConfig() {
+    const { cfg } = this;
+    /** @type {esbuild.BuildOptions} */
+    const opts = {
+      format: 'esm',
+      platform: 'node',
+      bundle: true,
+      outfile: path.relative(cfg.cwd, cfg.bundle),
+      external: [
+        '@aws-sdk/*',
+        // the following are imported by the universal adapter and are assumed to be available
+        './params.json',
+        'aws-sdk',
+        '@google-cloud/secret-manager',
+        '@google-cloud/storage',
+      ],
+      banner: {
+        js: [
+          'import { createRequire } from "module";',
+          'const require = createRequire(import.meta.url);',
+          '',
+        ].join('\n'),
+      },
+      entryPoints: [
+        cfg.adapterFile || path.resolve(__dirname, '..', 'template', 'node-index.mjs'),
+      ],
+      absWorkingDir: cfg.cwd,
+      plugins: [{
+        name: 'alias-main',
+        setup: (build) => {
+          build.onResolve({ filter: /^\.\/main\.js$/ }, () => ({ path: cfg.file }));
+          // use @adobe/helix-universal in the calling service, not ours
+          build.onResolve(
+            { filter: /^@adobe\/helix-universal$/ },
+            (args) => ({ path: path.resolve(cfg.cwd, 'node_modules', args.path, 'src', 'index.js') }),
```

Comment:
subtle point: when the bundler tries to resolve @adobe/helix-universal, and it adds its *own* version, it will also resolve other dependencies there (e.g. @adobe/fetch), which is unexpected.